internal/infrastructure/kubernetes/proxy/resource.go

```diff
@@ -204,6 +205,34 @@ func expectedProxyContainers(infra *ir.ProxyInfra,
 				SuccessThreshold: 1,
 				FailureThreshold: 3,
 			},
+			Lifecycle: &corev1.Lifecycle{
+				PreStop: &corev1.LifecycleHandler{
+					HTTPGet: &corev1.HTTPGetAction{
+						Path:   "/shutdown/ready",
+						Port:   intstr.FromInt32(envoy.ShutdownReadyPort),
+						Scheme: corev1.URISchemeHTTP,
+					},
+				},
+			},
+		},
+		{
+			Name:            "shutdown-manager",
```

Comment:
should this be a sidecar container ?
* this will ensure that `preStop` for `shutdown-manager` runs before `envoy`
* building off this
  * preStop exec `shutdown` performs a `touch` on the tmp file and exits
  * preStop exec `/shutdown/ready` performs the envoy shutdown procedure and blocks until complete
  * `/shutdown/ready` can return a non 200 error code immediately if tmp file doesnt exist

Reply:
Not sure of the shutdown sequencing as I don't see much about it in the [announcement blog post](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) or on the [docs page](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) other than indications that pod termination is not blocked by sidecar containers.

I could look into this if we are OK with dropping support for 1.28 where sidecars containers are feature gated only being enabled by default in 1.29. Not sure I'd recommend this though with 1.29 not yet being available as stable/GA on [AKS](https://learn.microsoft.com/en-us/azure/aks/supported-kubernetes-versions?tabs=azure-cli#aks-kubernetes-release-calendar) or [GKE](https://cloud.google.com/kubernetes-engine/docs/release-schedule#schedule-for-release-channels)

Reply:
ah good point, we cant rely on order yet, since sidecars just GA'd

then I propose `/shutdown/ready` sleeps for 1s (ensuring in case of termination, the other preStop had the chance to create the tmp file) before checking whether the tmp file exists or not. 
If it doesnt, return `404`, else run the shutdown steps and return `200`.
This eliminates the need for the `for` within the http handler.

Reply:
at a high level, big +1 with this approach

Reply:
> then I propose /shutdown/ready sleeps for 1s (ensuring in case of termination, the other preStop had the chance to create the tmp file) before checking whether the tmp file exists or not.
If it doesnt, return 404, else run the shutdown steps and return 200.

Not sure I follow. As soon as the `/shutdown/ready` request returns (no matter the response code) kubelet will immediately send sigterm to PID 1 (envoy).

Was the concern with the `for` loop that it could never return? I've modified this to ensure requests don't endlessly poll after the ready-timeout: 4a3b90f69d547cf36b1e8882a4d2f9ab5e6f9eb5 since `http.TimeoutHandler` of course wouldn't interrupt the processing just breaking the connection.

Reply:
the concerns I was trying to mitigate were
* using more CPU with the `for`
* allowing non kubelet clients to hoard the shutdown handler CPU

Since we cannot establish order between the `envoy` and `shutdown-manager` container, I was proposing we use a delay (`1s` / `2s`) to ensure that for the `preStop` case, we can be confident that the other `preStop` has signaled to us (using a file) that its time to shutdown.

For the non preStop case, it would just be a delay + return 404

Reply:
Updated in b3dcffd4b3b455cc391190a3c553b52295cd524b